clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/LambdaCachingAgent.java

```diff
@@ -212,4 +249,164 @@ public CacheResult loadData(ProviderCache providerCache) {
+
+    GetFunctionResult functionResult = null;
+    try {
+      functionResult = lambda.getFunction(new GetFunctionRequest().withFunctionName(functionName));
+    } catch (ResourceNotFoundException ex) {
+      log.info("Function {} Not exist", functionName);
+    }
+
+    Map<String, Object> attributes = new HashMap<String, Object>();
+    attributes.put("name", appName);
+    Map<String, Collection<String>> evictions = Collections.emptyMap();
+
+    Collection<String> existingFunctionRel = null;
+
+    CacheData application = providerCache.get(APPLICATIONS.ns, appKey);
+
+    if (null != application && null != application.getRelationships()) {
+      existingFunctionRel = application.getRelationships().get(LAMBDA_FUNCTIONS.ns);
+    }
+
+    Map<String, Collection<String>> relationships = new HashMap<String, Collection<String>>();
+
+    if (null != existingFunctionRel && !existingFunctionRel.isEmpty()) {
+      if (null == functionResult && existingFunctionRel.contains(functionKey)) {
+        existingFunctionRel.remove(functionKey);
+        evictions.put(LAMBDA_FUNCTIONS.ns, Collections.singletonList(functionKey));
+      } else {
+        existingFunctionRel.add(functionKey);
+      }
+
+    } else {
+      existingFunctionRel = Collections.singletonList(functionKey);
+    }
+
+    relationships.put(LAMBDA_FUNCTIONS.ns, existingFunctionRel);
+    DefaultCacheData cacheData = new DefaultCacheData(appKey, attributes, relationships);
+    DefaultCacheResult defaultCacheresults =
+        new DefaultCacheResult(
+            Collections.singletonMap(
+                com.netflix.spinnaker.clouddriver.core.provider.agent.Namespace.APPLICATIONS.ns,
+                Collections.singletonList(cacheData)));
+
+    return new OnDemandAgent.OnDemandResult(getAgentType(), defaultCacheresults, evictions);
+  }
+
+  @Override
+  public Collection<Map> pendingOnDemandRequests(ProviderCache providerCache) {
+    return null;
+  }
+
+  @Override
+  public String getOnDemandAgentType() {
+    return getAgentType() + "-OnDemand";
+  }
+
+  @Override
+  public OnDemandMetricsSupport getMetricsSupport() {
+    return metricsSupport;
+  }
+
+  private Boolean validKeys(Map<String, ? extends Object> data) {
+    return (data.containsKey("functionName")
+        && data.containsKey("credentials")
+        && data.containsKey("region"));
+  }
+
+  protected String combineAppDetail(String appName, String functionName) {
+    return Names.parseName(functionName).getApp().equals(appName)
+        ? functionName
+        : (appName + "-" + functionName);
+  }
+
+  private List<String> getTargetGroupNames(AWSLambda lambda, String functionName) {
+    List<String> targetGroupNames = new ArrayList<>();
+    try {
+      GetPolicyResult result =
+          lambda.getPolicy(new GetPolicyRequest().withFunctionName(functionName));
+      String json = result.getPolicy();
+      Policy policy = Policy.fromJson(json);
+      policy
```

Comment:
I believe this can be quite a bit less nested / simpler using stream/flatMap/filter instead of forEach/if

eg
```java
policy.getStatements
  .stream()
  .filter(statement -> statement.getEffect().toString().equals(Statement.Effect.Allow.toString()))
  .flatMap(statement -> statement.getActions.stream())
  .filter(action -> action.getActionName().equals("lambda:InvokeFunction"))
  ....
```

Reply:
I feel Steams impacts the performance and consume more heap. Mostly, I'll use streams for parallelism/sort.Let me know, your thoughts.

Reply:
I'm pretty happy to trade readability for memory/performance here: I'd need to see some concrete numbers on the cost of stream here to justify the indentation here.

Additionally, I'd avoid forEach if performance was a concern and use a for loop over the iterable instead of a forEach / function call